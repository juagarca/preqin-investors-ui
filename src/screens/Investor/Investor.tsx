import { useState } from "react";
import { useLocation, useParams } from "react-router-dom";
import { useQuery } from "@tanstack/react-query";

import { AssetClassesDropdown, CommitmentInfoTable } from "../../components";
import { fetchCommitmentInfo } from "../../utils/api";
import { AssetClass } from "../../types";

const Investor = () => {
  const location = useLocation();
  const { investorId } = useParams<{ investorId: string }>();
  const { firmName } = location.state || {};
  const [assetClassSelected, setAssetClassSelected] = useState<
    AssetClass | undefined
  >(undefined);

  const {
    data: commitmentInfoList,
    isLoading,
    error,
  } = useQuery({
    queryKey: ["commitmentInfo", assetClassSelected],
    queryFn: () => fetchCommitmentInfo(assetClassSelected!, investorId!),
    enabled: !!investorId && !!assetClassSelected,
  });

  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setAssetClassSelected(event.target.value as AssetClass);
  };

  if (!firmName) {
    return <div>Error!</div>;
  }

  return (
    <div>
      <h1>{firmName}</h1>
      <AssetClassesDropdown
        assetClassSelected={assetClassSelected}
        handleChange={handleChange}
      />
      {isLoading ? (
        <p>Loading... </p>
      ) : error ? (
        <p>Error loading commitment information</p>
      ) : commitmentInfoList && commitmentInfoList.length > 0 ? (
        <CommitmentInfoTable commitmentInfoList={commitmentInfoList} />
      ) : (
        commitmentInfoList && <p>No commitment information found</p>
      )}
    </div>
  );
};

export default Investor;
