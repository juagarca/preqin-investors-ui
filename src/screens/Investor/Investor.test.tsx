import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { useQuery } from "@tanstack/react-query";
import { MemoryRouter, Route, Routes } from "react-router-dom";

import Investor from "./Investor";

import ROUTES from "../../routes";
import investors from "../../utils/tests/mocks/investors.json";
import commitmentInfoList from "../../utils/tests/mocks/commitmentInfoList.json";

jest.mock("@tanstack/react-query");

describe("Investor component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = (state = { firmName: investors[0].firm_name }) => {
    const path = `${ROUTES.investors}/${investors[0].firm_id}`;

    return render(
      <MemoryRouter initialEntries={[{ pathname: path, state }]}>
        <Routes>
          <Route path={ROUTES.investor} element={<Investor />} />
        </Routes>
      </MemoryRouter>
    );
  };

  it("matches the snapshot", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: commitmentInfoList,
      isLoading: false,
      error: false,
    });

    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("successfully renders on initial render", () => {
    (useQuery as jest.Mock).mockReturnValue({
      enabled: false,
    });

    renderComponent();

    expect(screen.getByText(investors[0].firm_name)).toBeInTheDocument();
    expect(screen.getByText(/Pick an asset class/)).toBeInTheDocument();
    expect(screen.queryByText("ID")).not.toBeInTheDocument();
    expect(screen.queryByText("Currency")).not.toBeInTheDocument();
    expect(screen.queryByText("Amount")).not.toBeInTheDocument();
  });

  it("renders the CommitmentInfoTable when selecting an asset class", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: commitmentInfoList,
      isLoading: false,
      error: false,
    });

    renderComponent();

    const selectElement = screen.getByText(/Pick an asset class/);
    fireEvent.change(selectElement, { target: { value: "PE" } });

    expect(screen.getByText("ID")).toBeInTheDocument();
    expect(screen.getByText("Currency")).toBeInTheDocument();
    expect(screen.getByText("Amount")).toBeInTheDocument();

    commitmentInfoList.forEach((commitmentInfo) => {
      expect(screen.getByText(commitmentInfo.id)).toBeInTheDocument();
      expect(screen.getByText(commitmentInfo.currency)).toBeInTheDocument();
      expect(screen.getByText(commitmentInfo.amount)).toBeInTheDocument();
    });
  });

  it("renders message when no commitment information data found", () => {
    (useQuery as jest.Mock).mockReturnValue({
      data: [],
      isLoading: false,
      error: false,
    });

    renderComponent();

    expect(
      screen.getByText("No commitment information found")
    ).toBeInTheDocument();
  });

  it('renders loading when fetching"', () => {
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: true,
    });

    renderComponent();

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("renders error when a fetching error occurs", () => {
    (useQuery as jest.Mock).mockReturnValue({
      error: true,
    });

    renderComponent();

    expect(
      screen.getByText("Error loading commitment information")
    ).toBeInTheDocument();
  });

  it("renders error when accessing screen directly with url", () => {
    (useQuery as jest.Mock).mockReturnValue({
      error: true,
    });

    renderComponent({ firmName: "" });

    expect(screen.getByText("Error!")).toBeInTheDocument();
  });
});
