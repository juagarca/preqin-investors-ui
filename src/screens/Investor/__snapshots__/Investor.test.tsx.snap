// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investor component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Mjd Jedi fund
    </h1>
    <div
      class="investors-dropdown"
    >
      <select
        id="dropdown"
      >
        <option
          value=""
        >
          Pick an asset class to show its commitment information
        </option>
        <option
          value="pe"
        >
          Private Equity
        </option>
        <option
          value="pd"
        >
          Private Debt
        </option>
        <option
          value="re"
        >
          Real Estate
        </option>
        <option
          value="inf"
        >
          Infrastructure
        </option>
        <option
          value="nr"
        >
          Natural Resources
        </option>
        <option
          value="hf"
        >
          Hedge Funds
        </option>
      </select>
    </div>
    <table
      class="commitment-info-table"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            Currency
          </th>
          <th>
            Amount
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            34728
          </td>
          <td>
            HKD
          </td>
          <td>
            6M
          </td>
        </tr>
        <tr>
          <td>
            90806
          </td>
          <td>
            EUR
          </td>
          <td>
            20M
          </td>
        </tr>
        <tr>
          <td>
            90682
          </td>
          <td>
            GBP
          </td>
          <td>
            58M
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
